crates/cust/src/memory/device/device_slice.rs

```diff
@@ -426,8 +439,16 @@ pub trait DeviceSliceIndex<T: DeviceCopy> {
     /// # Safety
     ///
     /// The range must be in-bounds of the slice.
-    unsafe fn get_unchecked(self, slice: &DeviceSlice<T>) -> DeviceSlice<T>;
-    fn index(self, slice: &DeviceSlice<T>) -> DeviceSlice<T>;
+    unsafe fn get_unchecked(self, slice: &DeviceSlice<T>) -> &DeviceSlice<T>;
+    fn index(self, slice: &DeviceSlice<T>) -> &DeviceSlice<T>;
+
+    /// Indexes into this slice without checking if it is in-bounds.
+    ///
+    /// # Safety
+    ///
+    /// The range must be in-bounds of the slice.
+    unsafe fn get_unchecked_mut(self, slice: &mut DeviceSlice<T>) -> &mut DeviceSlice<T>;
+    fn index_mut(self, slice: &mut DeviceSlice<T>) -> &mut DeviceSlice<T>;
```

Comment:
The reason for reimplementing `SliceIndex` was that it was impossible to make a `&DeviceSlice<T>` because itd be returning a temporary value, reimplementing deviceslice as a DST should fix this and make `DeviceSliceIndex` useless, at that point we can implement it using normal `SliceIndex` (tho not sure how that behaves with a ZST slice). Or keeping `DeviceSliceIndex` but implementing `Index` and `IndexMut` again

Reply:
normal `SliceIndex` will be UB with `[T]` and won't change the pointer position correctly with `[()]` so we still need `DeviceSliceIndex`, but I'll impl `Index(Mut)`

Reply:
I see, i think we should just remove the index method we have already and reimplement the index traits. Not *too* excited about re-breaking the deviceslice API lmao but this will be a 0.4 thing.